netty/src/main/java/io/grpc/netty/NettyServerStream.java

```diff
@@ -203,7 +204,16 @@ public void deframeFailed(Throwable cause) {
       log.log(Level.WARNING, "Exception processing message", cause);
       Status status = Status.fromThrowable(cause);
       transportReportStatus(status);
-      handler.getWriteQueue().enqueue(new CancelServerStreamCommand(this, status), true);
+
+      Metadata trailers = new Metadata();
+      trailers.put(InternalStatus.CODE_KEY, status);
+      if (status.getDescription() != null) {
+        trailers.put(InternalStatus.MESSAGE_KEY, status.getDescription());
+      }
+      Http2Headers http2Trailers = Utils.convertTrailers(trailers, /* headersSent = */ false);
```

Comment:
We don't actually know whether headers have been sent. We need the handler to figure out that part.

CancelServerStreamCommand does have a status already, and it (properly) doesn't delay the closure, so there's no usage of closeStreamWhenDone(). Would it be better to put a boolean/enum on it for "prefer sending closure in trailers/metadata"? We can then in the handler check `steram.isHeadersSent()`. If we have sent the headers, flow control and such makes it more complicated, so we can just use RST there like today.

(Feel free to put an enum in the same class that is receiving it.)

Reply:
That seems totally reasonable, I'll take a stab at it.

Reply:
Done, and confirming that with this change my manual e2e test is still showing the right status:
```
WARNING: RPC failed: Status{code=RESOURCE_EXHAUSTED, description=gRPC message exceeds maximum size 4194304: 5000005, cause=null}
```